includes/Core/DI/DI_Container.php

```diff
@@ -0,0 +1,196 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\DI\DI_Container
+ *
+ * @package   Google\Site_Kit\Core\DI
+ * @copyright 2021 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\DI;
+
+use ArrayAccess;
+use Google\Site_Kit_Dependencies\Psr\Container\ContainerInterface;
+
+/**
+ * DI container class.
+ *
+ * @since n.e.x.t
+ */
+class DI_Container implements ContainerInterface, ArrayAccess {
+
+	/**
+	 * Definitions list.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	protected $definitions = array();
+
+	/**
+	 * Returns true if the container can return an entry for the given identifier.
+	 * Returns false otherwise.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to look for.
+	 * @return bool TRUE if the container can return an entry, otherwise FALSE.
+	 */
+	public function offsetExists( $id ) {
+		return $this->has( $id );
+	}
+
+	/**
+	 * Finds an entry of the container by its identifier and returns it.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to look for.
+	 * @return mixed Entry.
+	 */
+	public function offsetGet( $id ) {
+		return $this->get( $id );
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 * @param mixed  $entry Entry.
+	 */
+	public function offsetSet( $id, $entry ) {
+		$this->set( $id, $entry );
+	}
+
+	/**
+	 * Removes the entry.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to remove.
+	 */
+	public function offsetUnset( $id ) {
+		unset( $this->definitions[ $id ] );
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 * @param mixed  $entry Entry.
+	 * @return bool TRUE if the entry is added, otherwise FALSE.
+	 */
+	public function set( $id, $entry ) {
+		if ( ! empty( $this->definitions[ $id ]['is_protected'] ) ) {
+			return false;
+		}
+
+		$this->definitions[ $id ] = array(
+			'is_factory'   => false,
+			'is_protected' => false,
+			'is_service'   => is_callable( $entry ),
```

Comment:
This may not be entirely accurate, I think it's a bit too clever. For example, `$entry` will also be a callable for a factory, in which case it isn't a "service".

I'd prefer if the `set` method was more flexible (e.g. `$entry` would be an array with all the necessary data), and we add wrapper methods to simplify the more specific definitions, e.g. with a `set_service( $id, $create_func )` and `set_factory( $id, $create_func )` which then call the lower-level `set` accordingly. We could even introduce something like a `set_value( $id, $value )` for the most generic type which then would call `set( $id, array( 'entry' => $value )` - essentially `set` would always be the low-level method, and for each of the three types of entries there's a higher-level method to set them.

Reply:
Yes, i like this idea. Updated.